open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,639 @@
+  SPIEL_CHECK_EQ(player, CurrentPlayer());
+
+  if (!AllShipsPlaced_()) {
+    // If we are here, we still have some ships to place on the board.
+    //
+    // First, we find the first ship that hasn't been placed on the board yet.
+    const ShipPlacement ship_placement =
+        DeserializeShipPlacementAction_(action);
+    return ship_placement.ToString();
+  } else {
+    // In this case, the only thing the player can do is to shoot on a cell
+    const Shot shot = DeserializeShotAction_(action);
+    return shot.ToString();
+  }
+}
+
+std::string BattleshipState::ToString() const {
+  std::string state_str;
+  for (const auto& move : moves_) {
+    if (move.player == Player1) {
+      absl::StrAppend(&state_str, "/1:");
+    } else {
+      absl::StrAppend(&state_str, "/2:");
+    }
+    if (absl::holds_alternative<ShipPlacement>(move.action)) {
+      absl::StrAppend(&state_str,
+                      absl::get<ShipPlacement>(move.action).ToString());
+    } else {
+      SPIEL_DCHECK_TRUE(absl::holds_alternative<Shot>(move.action));
+      absl::StrAppend(&state_str, absl::get<Shot>(move.action).ToString());
+    }
+  }
+  return state_str;
+}
+
+bool BattleshipState::IsTerminal() const {
+  return CurrentPlayer() == kTerminalPlayerId;
+}
+
+std::vector<double> BattleshipState::Returns() const {
+  SPIEL_CHECK_TRUE(IsTerminal());
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+
+  // The description of the game in the header file contains more details
+  // about how the payoffs for the players are computed at the end of the
+  // game, as well as the meaning of the `loss_multiplier`.
+  const double loss_multiplier = conf.loss_multiplier;
+
+  double damage_pl1 = 0.0;
+  double damage_pl2 = 0.0;
+  for (const Ship& ship : conf.ships) {
+    if (DidShipSink_(ship, Player1)) damage_pl1 += ship.value;
+    if (DidShipSink_(ship, Player2)) damage_pl2 += ship.value;
+  }
+
+  return {damage_pl2 - loss_multiplier * damage_pl1,
+          damage_pl1 - loss_multiplier * damage_pl2};
+}
+
+std::unique_ptr<State> BattleshipState::Clone() const {
+  return std::make_unique<BattleshipState>(*this);
+}
+
+std::string BattleshipState::InformationStateString(Player player) const {
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+  const Player opponent = (player == Player1) ? Player2 : Player1;
+
+  // We will need to figure out whether each of the player's shots
+  // (i) hit the water, (ii) damaged but did not sink yet one of the opponent's
+  // ships, or (iii) damaged and sank one of the opponent's ships.
+  //
+  // To be able to figure that out, we will keep track of the damage that each
+  // of the opponent's ship has received so far. The vector `ship_damage`
+  // contains and updates this information as each player's shot is processed in
+  // order. Position i corresponds to the damage that the opponent's ship in
+  // position i of bs_game->configuration.ships has suffered.
```

Comment:
Does this make it a game of imperfect recall? I.e. you're forgetting the order that everything happened in. Though it seems like it's safe to do it in this game, as in it won't cause any problems.. can you confirm? 

If so, you should document it somewhere that this is explicitly the case, because we normally differentiate infostate from observation in this specific way. (Should not affect most algorithms using the infostate strings)

Reply:
Perhaps I misunderstood the difference between `InformationState` and `Observation`. In my mind, `InformationState` was an information set, while `Observation` could be anything, as long as it contains a subset of the information contained in `InformationState` (for example, the state of the board, without the history that led to it).

> you're forgetting the order that everything happened in
Here we are going in the order of the moves, so we are not really forgetting any order. The only difference between `InformationStateString(Player player)` and `state.ToString()` is that in the former the placement actions for the opponent are hidden, and shot outcomes (water / hit-but-not-sunk / hit-and-sunk) are added for `player`'s shots. Shot outcomes for the opponent are not needed, since those are a function of `player`'s placement and those are observed by `player`.

> we normally differentiate infostate from observation in this specific way
In the game of Battleship, I would say the only difference between the information state and an observation is only that in `Observation` the player does not distinguish between the order of moves, and only cares about the outcomes.


EDIT: Thinking back, I think I see where the misunderstanding is. I think you saw that I implemented `Observation` as simply a forward to `InformationState` and thought that therefore I was implementing `InformationState` by letting `player` forget about the specific order of the actions. In fact, the opposite happened: so far both remember the specific order of moves. So, going back to your question about whether the return value of `InformationState` defines a perfect-recall game, the answer is *yes*.

If you agree, I'll go ahead and implement `Observation` (this time for real, no forwarding to `InformationState`) by making sure that the order of the moves is forgotten by the player?

Reply:
Nevermind, misunderstanding on my part. This is indeed  correct!